### PR TITLE
Allow configuring serializer in file based queue

### PR DESCRIPTION
**Pull request by relud**
I'm interested in using this library, but I'm hesitant to use pickling.

Is it allowable for me to submit some PRs that make it possible to configure another serializer, such as json or msgpack?

Also, this PR should fix a couple bugs:
* if `path` exists, such as when resuming a queue after restart, then `self.protocol` would be `None`
* `_put()` wasn't actually using `self.protocol`, so it would fall back to version 0

### CODE REVIEW

**Comment by peter-wangxu** [comment]
Hi @relud 
You are welcome to make enhancements/bugfixes for persist-queue.

Your request looks fine for me, please fix the CI issue first.

Thanks
Peter

File: persistqueue/queue.py [hunk 1]
```diff
@@ -68,13 +68,24 @@ def __init__(self, path, maxsize=0, chunksize=100, tempdir=None):
...
+
+        The serialize parameter controls how enqueued data is serialized. It
+        must be a callable that accepts two args. The first arg is the data
+        to serialize, and the second arg is a byte mode file object the
+        serialization should be written to. It must serialize data so that
+        deserialize can read one object from the file at a time.
+
+        The deserialize parameter controls how enqueued data is deserialized.
+        It must be a callable that accepts one arg, a byte mode file object,
+        and reads one object that is returned.
```
Comment (relud) [inline persistqueue/queue.py]: I went to implement msgpack and json serializers, and found that it was much easier to configure with callables than a single `serializer` with `load` and `dump` methods. let me know if you would prefer a different implementation.

Comment (peter-wangxu) [reply persistqueue/queue.py]: I have to say, I prefer the latter one, it is following the OOP, and easier to extend.
And do keep the backward compatibility.

Thanks
Peter

**Comment by peter-wangxu** [comment]
Do you have a change to run local unittest on Linux? The circle CI(for linux) was not triggered somehow

File: persistqueue/queue.py (hunk 1 above)
Comment (relud) [reply persistqueue/queue.py]: okay. I'll restructure it to pass a serializer again on Monday (PDT timezone).

**Comment by relud** [comment]
i don't think i did anything that would disable the linux CI. maybe it is because it is a pull request from a fork?

**Comment by peter-wangxu** [comment]
> i don't think i did anything that would disable the linux CI. maybe it is because it is a pull request from a fork?

my typo(change->chance) misled you:) Yes, it's Circle CI issue. I was wondering if you could run the existing tests via `tox` on Linux, otherwise the Linux run could only be run after PR merged.

File: persistqueue/queue.py (hunk 1 above)
Comment (peter-wangxu) [reply persistqueue/queue.py]: 👍

**Comment by relud** [comment]
I enabled circleci on my fork, and restructured to pass around a serializer object.

**Comment by peter-wangxu** [comment]
LGTM，If no further code change, I am happy to merge it.

**Comment by relud** [comment]
yep, this is done!

**Pull request merged**